ferris_elf/picoseconds.py

```diff
@@ -0,0 +1,119 @@
+import operator as op
+from typing import Optional, Self, Callable
+from warnings import warn
+
+
+def format_picos(ts: float | int) -> str:
+    timestamp = float(ts)
+
+    base = "ps"
+    scalar: list[tuple[str, int]] = [
+        ("ns", 1000),
+        ("µs", 1000),
+        ("ms", 1000),
+        ("s", 1000),
+        ("m", 60),
+        ("h", 60),
+    ]
+
+    for name, offset in scalar:
+        if timestamp > offset:
+            timestamp /= offset
+            base = name
+        else:
+            break
+
+    # remove any trailing stuff
+    timestamp = round(timestamp, 2)
+
+    # fun awesome fact, int decays to float in typeck directly,
+    # but is_integer is only implemented on float so if we didn't
+    # explicitly call timestamp = float(ts), this could break
+    # and typeck would be ok with it :D
+    if timestamp.is_integer():
+        return f"{timestamp:.0f}{base}"
+    else:
+        return f"{timestamp:.2f}{base}"
+
+
+# Base class interferes with our typing
+def mk_binop(
+    f: Callable[[int, int], int],
+) -> Callable[["Picoseconds | int", int], "Picoseconds | int"]:
+    return lambda self, other: Picoseconds.from_picos(f(int(self), other))
+
+
+def mk_prefixop(f: Callable[[int], int]) -> Callable[["Picoseconds | int"], "Picoseconds | int"]:
```

Comment:
why are we annotating the callable as returning `Picoseconds | int`? it always returns `Picoseconds`

Reply:
Mypy complains about compatibility with the base class `int` if I don't. Could have sworn I left a comment about that somewhere. Turns out, once you dive into the details around typing, it just doesn't make a lot of sense to inherit from int.